incubator/hnc/pkg/validators/object.go

```diff
@@ -0,0 +1,117 @@
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="rbac.authorization.k8s.io",resources=rols,verbs=create;update,versions=v1,name=roles.rbac.authorization.k8s.io
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=create;update,versions=v1,name=rolebindings.rbac.authorization.k8s.io
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="networking.k8s.io",resources=networkpolicies,verbs=create;update,versions=v1,name=networkpolicies.networking.k8s.io
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=resourcequotas,verbs=create;update,versions=v1,name=resourcesquotas
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=limitranges,verbs=create;update,versions=v1,name=limitranges
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=configmaps,verbs=create;update,versions=v1,name=configmaps
+
+type Object struct {
+	Log     logr.Logger
+	Forest  *forest.Forest
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+func (o *Object) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := o.Log
+	inst := &unstructured.Unstructured{}
+	if err := o.decoder.Decode(req, inst); err != nil {
+		log.Error(err, "Couldn't decode request")
+		return deny(metav1.StatusReasonBadRequest, err.Error())
+	}
+	log = log.WithValues("object", inst.GetName())
+
+	// oldInst is the existing object. It is null for CREATE and CONNECT.
+	oldInst := &unstructured.Unstructured{}
+	if err := o.decoder.DecodeRaw(req.OldObject, oldInst); err != nil {
+		log.Error(err, "Couldn't decode RAW request")
+		return deny(metav1.StatusReasonBadRequest, err.Error())
+	}
+
+	if isHNCServiceAccount(&req.AdmissionRequest.UserInfo) {
+		log.Info("Allowed change by HNC SA")
+		return allow("HNC SA")
+	}
+
+	resp := o.handle(ctx, log, inst, oldInst)
+	log.Info("Handled", "allowed", resp.Allowed, "code", resp.Result.Code, "reason", resp.Result.Reason, "message", resp.Result.Message)
+	return resp
+}
+
+// handle implements the non-webhook-y businesss logic of this validator, allowing it to be more
+// easily unit tested (ie without constructing an admission.Request, setting up user infos, etc).
+func (o *Object) handle(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured, oldInst *unstructured.Unstructured) admission.Response {
+	oldValue, oldOK := getLabel(oldInst, api.LabelInheritedFrom)
+	newValue, newOK := getLabel(inst, api.LabelInheritedFrom)
+	// If old object holds the label but the new one doesn't, reject it. Vice versa.
+	if oldOK != newOK {
+		verb := "add"
+		if !newOK {
+			verb = "remove"
+		}
+		return deny(metav1.StatusReasonUnauthorized, "Users should not"+verb+"label"+api.LabelInheritedFrom)
```

Comment:
"Unauthorized" requests can be retried with different authorization, but that isn't what we want here - _no_ user should be able to set this. So I'd go with "Forbidden" (403) instead.

Reply:
Also, you need spaces in the message, I think:

```
"Users should not "+verb+" the label "+api.LabelInheritedFrom
                 ^        ^^^^^     ^
```

Reply:
Done.